### PR TITLE
Adds endpoints resource

### PR DESCRIPTION
**Pull request by JosephSalisbury**
Towards https://github.com/giantswarm/giantswarm/issues/2372

### CODE REVIEW

File: service/controller/v1/resource/endpoints/resource.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+func toEndpoints(v interface{}) (*apiv1.Endpoints, error) {
+	if v == nil {
+		return nil, nil
+	}
+
+	endpoints, ok := v.(*apiv1.Endpoints)
+	if !ok {
+		return nil, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", &apiv1.Endpoints{}, v)
```
Comment (kopiczko) [inline service/controller/v1/resource/endpoints/resource.go]: I'd like to replace that with `return nil, microerror.Maskf(wrongTypeError, "expected '%T', got '%T'", endpoints, v)`. That way we can be sure the printed type correct.

File: service/controller/v1/resource/endpoints/create.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+	desiredEndpoints, err := toEndpoints(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var endpointsToCreate *apiv1.Endpoints
+	if currentEndpoints == nil || desiredEndpoints.Name != currentEndpoints.Name {
```
Comment (kopiczko) [inline service/controller/v1/resource/endpoints/create.go]: Again I don't get this name comparison.

File: service/controller/v1/resource/endpoints/current.go [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+	namespace := key.ClusterNamespace(customObject)
+
+	// Lookup the current state of the endpoints.
+	var endpoints *apiv1.Endpoints
+	{
+		manifest, err := r.k8sClient.CoreV1().Endpoints(namespace).Get(masterEndpointsName, apismetav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find the master endpoints in the Kubernetes API")
+			// fall through
```
Comment (kopiczko) [inline service/controller/v1/resource/endpoints/current.go]: This is probably copy/paste but I would `return nil, nil` here explicitly.

File: service/controller/v1/resource/endpoints/desired.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	endpoints := &v1.Endpoints{
+		ObjectMeta: apismetav1.ObjectMeta{
+			Name:      "master",
+			Namespace: key.ClusterID(customObject),
+			Labels: map[string]string{
+				"app":                        "master",
+				"cluster":                    key.ClusterID(customObject),
+				"customer":                   key.ClusterCustomer(customObject),
+				"giantswarm.io/cluster":      key.ClusterID(customObject),
+				"giantswarm.io/organization": key.ClusterCustomer(customObject),
```
Comment (kopiczko) [inline service/controller/v1/resource/endpoints/desired.go]: I'd like to start adding `giantswarm.io/managed-by` label to the new manifests.

File: service/controller/v1/resource/endpoints/nic.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	privateIP := ""
+	for _, ipConfiguration := range *networkInterface.IPConfigurations {
+		if *ipConfiguration.PrivateIPAddress != "" {
+			privateIP = *ipConfiguration.PrivateIPAddress
+		}
+	}
```
Comment (kopiczko) [inline service/controller/v1/resource/endpoints/nic.go]: I'd like to assert here there is only one IP configuration.

File: service/controller/v1/resource/endpoints/nic.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	networkInterface, err := interfacesClient.Get(
+		context.Background(),
+		resourceGroupName,
+		networkInterfaceName,
+		expands,
+	)
```
Comment (kopiczko) [inline service/controller/v1/resource/endpoints/nic.go]: Nice

File: service/controller/v1/resource/endpoints/desired.go (hunk 4 above)
Comment (JosephSalisbury) [reply service/controller/v1/resource/endpoints/desired.go]: Done

File: service/controller/v1/resource/endpoints/create.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
+package endpoints
+
+import (
+	"context"
+
+	apiv1 "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply service/controller/v1/resource/endpoints/create.go]: Ok, nevermind :D

**Pull request merged**